add AudioDiffusionPipeline and LatentAudioDiffusionPipeline

teticio opened the pull request:
I have added `AudioDiffusionPipeline` and `LatentAudioDiffusionPipeline` which I intend to migrate from `https://github.com/teticio/audio-diffusion`. I have added them to the main `src` as opposed to the community pipelines due to the inheritance of `LatentAudioDiffusionPipeline` from `AudioDiffusionPipeline`, which cannot be done in a single pipeline file, as well as the fact that the `Mel` class is needed to convert from audio to images and vice versa. It might make sense to move the `Mel` class somewhere more central, as it could be used by other pipelines.

teticio commented:
@patrickvonplaten @Vaibhavs10 I'd be very grateful if you could have a look at this. Some of the tests are failing, but I don't think this is due to me.

teticio commented:
At least it be of the tests is failing due to the new dependency 'librosa'. I added this to 'setup.py' - is there anything else I need to do?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_1334). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten commented on src/diffusers/__init__.py:
```diff
@@ -30,12 +30,15 @@
+        LatentAudioDiffusionPipeline,
         LDMPipeline,
         LDMSuperResolutionPipeline,
+        Mel,
```
Could we remove this from the general `__init__.py` function? -> I don't think one would use `"Mel"` witouth the pipelines no? :-)

teticio replied:
A mel instance is a parameter to the pipeline and it is useful for creating the dataset and training the model. I agree that it will probably only be used in conjunction with the pipelines, so as long as you can import it from diffusers.pipelines that should be Ok. Is that what you mean? Thanks!

patrickvonplaten commented on docs/source/using-diffusers/audio.mdx:
```diff
@@ -12,5 +12,5 @@ specific language governing permissions and limitations under the License.
-The [`DanceDiffusionPipeline`] can be used to generate audio rapidly!
-More coming soon!
\ No newline at end of file
+[`DanceDiffusionPipeline`],  [`AudioDiffusionPipeline`] and [`LatentAudioDiffusionPipeline`] can be used to generate
```
Cool!

patrickvonplaten commented on docs/source/api/pipelines/audio_diffusion.mdx:
```diff
@@ -0,0 +1,112 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# Audio Diffusion
```
Cool docs!

patrickvonplaten commented on docker/diffusers-pytorch-cpu/Dockerfile:
```diff
@@ -36,6 +37,7 @@ RUN python3 -m pip install --no-cache-dir --upgrade pip && \
         tensorboard \
-        transformers
+        transformers \
+        librosa
```
Ok for me @anton-l what do you think?

teticio replied:
Thanks for reviewing @patrickvonplaten !

patrickvonplaten commented on tests/pipelines/audio_diffusion/test_audio_diffusion.py:
```diff
@@ -0,0 +1,159 @@
+    def test_audio_diffusion(self):
+        device = torch_device
+
+        mel = Mel()
```
Should we maybe create this automatically directly in the pipeline? It might be a bit more user-friendly?

patrickvonplaten commented on tests/pipelines/audio_diffusion/test_audio_diffusion.py:
```diff
@@ -0,0 +1,159 @@
+import numpy as np
+import torch
+
+from diffusers import (
```
Very nice tests!

patrickvonplaten commented on src/diffusers/pipelines/audio_diffusion/pipeline_audio_diffusion.py:
```diff
@@ -0,0 +1,216 @@
+    @torch.no_grad()
+    def __call__(
+        self,
+        mel: Mel,
```
It's a bit weird to me that one has to pass a class here for `__call__` -> wouldn't it be better to just do this inside the pipeline?

patrickvonplaten reviewed (commented):
Hey @teticio,

This looks generally very nice to me and I think we can merge this soon :-) 
One big thing we need to change is to make `librosa` an optional dependency. 
You can do this as follows:
- Add some "is_librosa_available()" logic here: https://github.com/huggingface/diffusers/blob/ab1f01e63415b63937736299d3a770554c83987e/src/diffusers/utils/import_utils.py#L129

And then import the pipeline only if `librosa` is available. Maybe similar to how we do it for the LMSScheduler here: https://github.com/huggingface/diffusers/blob/ab1f01e63415b63937736299d3a770554c83987e/src/diffusers/__init__.py#L61 => e.g. only import your pipelines if librosa is available :-) 


Also I would maybe not accept an "empyt" `Mel()` class as an input to the call function, IMO that's a bit unintuitive design-wise - could we maybe just better create this inside the call method? Wdyt? 

Finally, let's maybe not add `Mel` to the public init as I don't think anybody would import just the `Mel` class no? 


Overall, great work though! Very happy to soon have a second audio diffusion model :heart_eyes:

patrickvonplaten commented on setup.py:
```diff
@@ -187,7 +188,8 @@ def run(self):
     "torchvision",
-    "transformers"
+    "transformers",
+    "librosa"
```
This is fine for me @anton-l wdyt?

patrickvonplaten commented:
@anton-l could you quickly check the docker file changes here regarding the new `librosa` dependency?  No need to review the whole PR :-)

teticio commented:
> Hey @teticio,
> 
> 
> 
> This looks generally very nice to me and I think we can merge this soon :-) 
> 
> One big thing we need to change is to make `librosa` an optional dependency. 
> 
> You can do this as follows:
> 
> - Add some "is_librosa_available()" logic here: https://github.com/huggingface/diffusers/blob/ab1f01e63415b63937736299d3a770554c83987e/src/diffusers/utils/import_utils.py#L129
> 
> 
> 
> And then import the pipeline only if `librosa` is available. Maybe similar to how we do it for the LMSScheduler here: https://github.com/huggingface/diffusers/blob/ab1f01e63415b63937736299d3a770554c83987e/src/diffusers/__init__.py#L61 => e.g. only import your pipelines if librosa is available :-) 
> 
> 

Great, thanks for the tip. I'll do that tomorrow.

> 
> 
> 
> Also I would maybe not accept an "empyt" `Mel()` class as an input to the call function, IMO that's a bit unintuitive design-wise - could we maybe just better create this inside the call method? Wdyt? 
> 
> 

The Mel class encapsulates a few parameters (like hop length and so on) which I decided against adding to the model configs, so as not to pollute things. However, I can default the parameter so that it creates a Mel with the default parameters. I initially decided against this because I wanted to make it explicit that these parameters need setting. Do you think the best thing is to add the parameters like hop length etc to the pipeline __call__ with suitable defaults that are then used to create the mel object? Happy to follow your guidance here.

> 
> Finally, let's maybe not add `Mel` to the public init as I don't think anybody would import just the `Mel` class no? 
> 
> 
So I think the Mel class will be imported separately from the pipeline (for dataset creation and training). Shall I just make it importable from diffusers.pipelines and not from diffusers?
> 
> 
> 
> Overall, great work though! Very happy to soon have a second audio diffusion model :heart_eyes: 

Thank you! Very excited to add my 2 cents to this excellent repo!

teticio commented:
...or maybe it makes more sense to pass the Mel objct (or the parameters needed to as kwargs in the constructor instead of the __call__ method. Let me know what you think.

patrickvonplaten commented:
Hey @teticio,

Sorry I think the commit history got messed up :-/ 
Could you maybe try to fix it or just open a new PR if it's easier :sweat_smile: 

Yes, definitely fine for me to pass `Mel()` to the init method of the model

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_1334). All of your documentation changes will be reflected on that endpoint.

teticio commented:
Hi @patrickvonplaten 

> Hey @teticio,
> 
> Sorry I think the commit history got messed up :-/ Could you maybe try to fix it or just open a new PR if it's easier 😅
> 

Yeah, sorry about that - not sure what I did wrong there, but it should be OK now.

> Yes, definitely fine for me to pass `Mel()` to the init method of the model

So I took a different tack. I thought it would make more sense for Mel to be set up in the constructor of the pipeline. For that to be possible, it needs to be a module. So I moved it to models and derived it from ConfigMixin. For it to be able to be load_from_pretrained, I added Mel to the LOADABLE_CLASSES. The thinking is that this module could be replaced by some other audio<->image transformation in a composable way. For example, a neural one instead of the mel spectrogram. Arguably, there should be a base class (e.g., Audio2Image or something) from which Mel is derived, but I thought it might make more sense to refactor that if and when an alternative is implemented.

The great advantage of doing it this way is that the Mel object is guaranteed to be consistent with the rest of the model; before the user had to make sure that he / she was using the configuration used to train the model in the first place.

For this to work, I have to update the models uploaded to HF hub to include the Mel config. So currently the 'slow' test will fail. However, the colab notebook linked in the documentation is currently pointing to a modified version of the models in HF hub, so you can try it out with a pre-trained model there.

teticio commented:
> => e.g. only import your pipelines if librosa is available :-)

Done!

teticio commented:
@patrickvonplaten As soon as you confirm that this approach is acceptable (making Mel a module in the pipeline), I will make the corresponding changes in my current repo (from which I am migrating) and update the model repos accordingly, so that there can be a seamless switch over. Here is an example of how it will look in the model config: https://huggingface.co/teticio/latent-audio-diffusion-ddim-256-new/blob/main/mel/mel_config.json

teticio commented:
Ah, one last thing. If you don't like Mel being a LOADABLE_CLASS, maybe we could make ConfigMixin  one instead. This would allow config of classes without model weights other than Schedulers to be loaded and saved.

teticio commented:
@patrickvonplaten I put you back as reviewer to check the changes I made to Mel ^. I didn't realize it would remove anton-I

teticio commented:
Sorry. I dk wtf happened with my repo. I am going to start a new PR from scratch
